File: lesson_10/peaks.py
def solution(a):
  def next_divisor(i):
    i += 1
    while i <= n:
      if n % i == 0:
        return i
      else:
        i += 1
    return 0

  def find_blocks(i):
    for j in xrange(n/i):
      cnt = 0
      for k in xrange(i):
        cnt += b[j * i + k]
      if cnt == 0:
        return 0
    return n/i

  n = len(a)

  b = [0] * n
  for i in xrange(1, n-1):
    if a[i] - a[i-1] > 0 and a[i+1] - a[i] < 0:
      b [i] = 1

  ret = 0
  i = next_divisor(0)
  while i != 0:
    ret = find_blocks(i)
    if ret != 0:
      return ret
    i = next_divisor(i)
  return 0

File: 10_prime_and_composite_numbers/peaks.cpp
#include <vector>

using namespace std;

vector<int> peak_cnts(vector<int> &a) {
  int n = a.size();
  int cnt = 0;
  vector<int> ret(n + 1, 0);
  for (int i = 1; i < n - 1; i++) {
    if (a[i] > a[i - 1] && a[i] > a[i + 1]) {
      cnt++;
    }
    ret[i + 1] = cnt;
  }
  ret[n] = ret[n - 1];
  return ret;
}

bool satisfiable(vector<int> &pcnts, int k) {
  int n = pcnts.size() - 1;
  if (n % k != 0) {
    return 0;
  }
  for (int i = 0; i < n / k; i++) {
    if ((pcnts[(i + 1) * k] - pcnts[i * k]) == 0) {
      return 0;
    }
  }
  return 1;
}

int solution(vector<int> &a) {
  int n = a.size();
  vector<int> pcnts = peak_cnts(a);
  for (int k = 1; k <= n; k++) {
    if (satisfiable(pcnts, k)) {
      return n / k;
    }
  }
  return 0;
}
